worlds/faxanadu/__init__.py

```diff
@@ -0,0 +1,206 @@
+            self.put_wingboot_in_shop(shops, region)
+
+        return wingboots_count
+
+
+    def create_items(self):
+        itempool: List[FaxanaduItem] = []
+
+        # Prefill red potions in shops if option is set
+        self.prefill_shop_red_potions()
+
+        # Prefill wingboots in shops
+        wingboots_in_shop_count = self.prefill_shop_wingboots()
+
+        # Create the item pool, excluding fillers.
+        prefilled_count = self.red_potion_in_shop_count + wingboots_in_shop_count
+        for item in Items.items:
+            # Ignore pendant if turned off
+            if item.name == "Pendant" and not self.options.include_pendant.value:
+                continue
+
+            # ignore fillers for now, we will fill them later
+            if item.classification in [ItemClassification.filler, ItemClassification.trap] and \
+               item.progression_count == 0:
+                continue
+
+            prefill_loc = None
+            if item.prefill_location:
+                prefill_loc = self.multiworld.get_location(item.prefill_location, self.player)
+
+            # if require dragon slayer is turned on, we need progressive shields to be progression
+            item_classification = item.classification
+            if self.options.require_dragon_slayer.value and item.name == "Progressive Shield":
+                item_classification = ItemClassification.progression
+
+            if prefill_loc:
+                prefill_loc.place_locked_item(FaxanaduItem(item.name, item_classification, item.id, self.player))
+                prefilled_count += 1
+            else:
+                for i in range(item.count - item.progression_count):
+                    itempool.append(FaxanaduItem(item.name, item_classification, item.id, self.player))
+                for i in range(item.progression_count):
+                    itempool.append(FaxanaduItem(item.name, ItemClassification.progression, item.id, self.player))
+
+        # Randomly add fillers to the pool with ratios based on og game occurence counts.
+        filler_count = len(Locations.locations) - len(itempool) - prefilled_count
+        for i in range(filler_count):
+            itempool.append(self.create_item(self.get_filler_item_name()))
+                
+        self.multiworld.itempool += itempool
+
+
+    def get_filler_item_name(self) -> str:
+        # Addadd fillers to the pool with ratios based on og game occurence counts.
+        filler_ratios = {
+            item.name: item.count
+            for item in Items.items
+            if item.classification in [ItemClassification.filler, ItemClassification.trap]
+        }
+
+        # If red potions are locked in shops, remove the count from the ratio.
+        filler_ratios["Red Potion"] -= self.red_potion_in_shop_count
+
+        # Remove poisons if not desired
+        if not self.options.include_poisons.value:
+            filler_ratios["Poison"] = 0
```

Comment:
since these ratios are going to be static for any given world instance and get_filler_item_name() will be called multiple times, this instantiating should probably be in generate_early() or similar

Reply:
Ya it was cached before, but I moved them into this function as per another review. It's also dependent on how many `red_potion_in_shop_count`, so part of the randomization is done. 🤔

Reply:
But yea I'll cache the filler ratios in a member variable. this function is expensive

Reply:
haha, now that `red_potion_in_shop_count` is only used in create_items() you can make it into a local return like `wingboots_in_shop_count`